src/documentation/io/testing.md

```diff
@@ -108,11 +108,464 @@ This allows the I/O transform's unit tests to act as if they have a perfect netw
+*   **Your test should use pipeline options to receive connection information.**
+    *   For Java, there is a shared pipeline options object in the io/common directory. This means that if there are two tests for the same data store (e.g. for `Elasticsearch` and the `HadoopInputFormatIO` tests), those tests share the same pipeline options.
+*   **Generate test data programmatically and parameterize the amount of data used for testing.**
+    *   For Java, `CountingInput` + `TestRow` can be combined to generate deterministic test data at any scale.
+*   **Use a write then read style for your tests.**
+    *   In a single `Test`, run a pipeline to do a write using your I/O transform, then run another pipeline to do a read using your I/O transform.
+    *   The only verification of the data should be the result from the read. Don't validate the data written to the database in any other way.
+    *   Validate the actual contents of all rows in an efficient manner. An easy way to do this is by taking a hash of the rows and combining them. `HashingFn` can help make this simple, and `TestRow` has pre-computed hashes.
+    *   For easy debugging, use `PAssert`'s `containsInAnyOrder` to validate the contents of a subset of all rows.
+*   **Tests should assume they may be run multiple times and/or simultaneously on the same database instance.**
+    *   Clean up test data: do this in an `@afterClass` to ensure it runs.
+    *   Use unique table names per run (timestamps are an easy way to do this) and per-method where appropriate.
+
+An end to end example of these principles can be found in [JdbcIOIT](https://github.com/ssisk/beam/blob/jdbc-it-perf/sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOIT.java).
+
+
+#### Kubernetes scripts {#kubernetes-scripts}
+
+As discussed in [Integration tests, data stores, and Kubernetes](#integration-tests-data-stores-and-kubernetes), to have your tests run on Beam's continuous integration server, you'll need to implement a Kubernetes script that creates an instance of your data store.
+
+If you would like help with this or have other questions, contact the Beam dev@ mailing list and the community may be able to assist you.
+
+Guidelines for creating a Beam data store Kubernetes script:
+1.  **You must only provide access to the data store instance via a `NodePort` service.**
+    *   This is a requirement for security, since it means that only the local network has access to the data store. This is particularly important since many data stores don't have security on by default, and even if they do, their passwords will be checked in to our public Github repo.
+1.  **Convention: define two Kubernetes scripts.**
+    *   This is the best known way to implement item #1.
+    *   The first script will be the core script run on the Beam CI servers and should contain the main datastore instance script (`StatefulSet`) plus a `NodePort` service exposing the data store.
```

Comment:
The first script will ~~be the core script run on the Beam CI servers and should~~ contain (the text does not add a lot)